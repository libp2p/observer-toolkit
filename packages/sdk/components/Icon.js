import React from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import icons from '../theme/icons'

const Container = styled.span`
  display: inline-block;
  vertical-align: middle;
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  border-radius: ${({ size }) => size}px;

  ${({ theme, active }) =>
    active
      ? `
      color: ${theme.color('tertiary', 'mid')};
      background: ${theme.color('light', 'light')};
    `
      : `
      background: ${theme.color('light', 'light', 0.5)};
    `};
  ${({ theme, offset }) => offset && `margin-right: -${theme.spacing(2)};`}
  ${({ onClick, theme }) =>
    onClick &&
    `
    cursor: pointer;
    &:hover {
      backgroundColor: ${theme.color('light', 'mid')};
    }
  `}
`

function Icon({ type, onClick, active, size = 20 }) {
  const IconSvg = icons[type]

  return (
    <Container onClick={onClick} active={active} size={size}>
      <IconSvg />
    </Container>
  )
}

Icon.propTypes = {
  type: T.string.isRequired,
  onClick: T.func,
  active: T.bool,
  offset: T.bool,
  size: T.number,
}

export default Icon
