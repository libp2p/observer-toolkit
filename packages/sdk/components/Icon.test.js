import React from 'react'
import { fireEvent } from '@testing-library/react'
import { catchErrorSilently, renderWithTheme } from '@libp2p-observer/testing'
import Icon from './Icon'

describe('Icon', () => {
  it('throws if trying to render an invalid icon', () => {
    expect(
      catchErrorSilently(() =>
        renderWithTheme(<Icon type="invalid icon type" />)
      )
    ).toBeInstanceOf(Error)
  })

  it("renders 'edit' icon as expected", () => {
    const { asFragment } = renderWithTheme(<Icon type="edit" disabled={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('acts as button calling onClick if clicked', () => {
    const mockFn = jest.fn()

    const { getByRole } = renderWithTheme(
      <Icon type="edit" disabled={false} onClick={mockFn} />
    )
    fireEvent.click(getByRole('button'))
    expect(mockFn).toBeCalled()
  })

  it("doesn't act as button or call onClick if disabled", () => {
    const mockFn = jest.fn()

    const { getByRole, queryByRole } = renderWithTheme(
      <Icon type="edit" disabled={true} onClick={mockFn} />
    )
    expect(queryByRole('button')).toBeNull()

    fireEvent.click(getByRole('img'))
    expect(mockFn).not.toBeCalled()
  })
})
