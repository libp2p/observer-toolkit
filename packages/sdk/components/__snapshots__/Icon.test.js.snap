// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders 'edit' icon as expected 1`] = `
<DocumentFragment>
  .c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 2em;
  height: 2em;
  color: rgb(255,255,255);
}

.c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

<div
    class="c0"
  >
    <span
      class="c1"
      disabled=""
      role="img"
    >
      <svg
        viewBox="0 0 96 96"
      >
        <defs>
          <style>
            .edit_svg__cls-1{fill:none;stroke:currentColor;stroke-miterlimit:10;stroke-width:4px}
          </style>
        </defs>
        <g
          data-name="Layer 13"
          id="edit_svg__Layer_13"
        >
          <path
            class="edit_svg__cls-1"
            d="M23.63 57.63l14.74 14.74"
          />
          <path
            d="M69.72 18.84a7.5 7.5 0 015.1 2.34c2.24 2.23 3 5.45 1.74 7.28L73.05 32 31.49 73.6l-12.09 3 3-12.09L63.94 23l3.6-3.54a4 4 0 012.18-.6m0-4a7.71 7.71 0 00-4.78 1.55l-3.8 3.73-42.35 42.33-3.88 15.6a2.49 2.49 0 002.41 3.09 2.6 2.6 0 00.61-.07l15.6-3.88 42.35-42.35 3.73-3.8c2.64-3.4 1.86-8.9-2-12.71a11.44 11.44 0 00-7.93-3.51z"
            fill="currentColor"
          />
          <path
            class="edit_svg__cls-1"
            d="M60.84 24.74l10.64 10.64"
          />
        </g>
      </svg>
    </span>
  </div>
</DocumentFragment>
`;
